linkd-lib/src/socket_activation/unix.rs

```diff
@@ -0,0 +1,67 @@
+// Copyright © 2021 The Radicle Link Contributors
+//
+// This file is part of radicle-link, distributed under the GPLv3 with Radicle
+// Linking Exception. For full terms see the included LICENSE file.
+
+//! Implementation of the systemd socket activation protocol.
+//! http://0pointer.de/blog/projects/socket-activation.html
+//!
+//! TODO
+//! * support FDs beyond 3
+//! * support mapping from listen names
+
+use std::{
+    env,
+    os::unix::{io::RawFd, net::UnixListener, prelude::FromRawFd},
+};
+
+use anyhow::{bail, Result};
+use nix::{
+    fcntl::{fcntl, FcntlArg::F_SETFD, FdFlag},
+    sys::socket::SockAddr,
+    unistd::Pid,
+};
+
+/// Environemnt variable which carries the amount of file descriptors passed
+/// down.
+const LISTEN_FDS: &str = "LISTEN_FDS";
+/// Environment variable containing colon-separated list of names corresponding
+/// to the `FileDescriptorName` option in the service file.
+const _LISTEN_NAMES: &str = "LISTEN_NAMES";
+/// Environemnt variable when present should match PID of the current process.
+const LISTEN_PID: &str = "LISTEN_PID";
+
+pub fn env() -> Result<Option<UnixListener>> {
+    // TODO(xla): Enable usage of more than the first fd. For now the assumption
```

Comment:
What would be the purpose of using multiple file descriptors?

Reply:
Systemd unit files support the declaration of N named sockets, these are then used by the binary at own discretion. Use-cases include but not limited to offering multiple APIs or offer telemetry on an isolated socket.

> sd_listen_fds_with_names() is like sd_listen_fds(), but optionally also returns an array of strings with identification names for the passed file descriptors

From https://www.freedesktop.org/software/systemd/man/sd_listen_fds.html

Reply:
Gotcha, thanks.